library/ssl_tls.c

```diff
@@ -2855,21 +2871,91 @@ int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
             ssl_swap_epochs( ssl );
         }
 
-        memcpy( ssl->out_msg, cur->p, cur->len );
-        ssl->out_msglen = cur->len;
-        ssl->out_msgtype = cur->type;
+        /* CCS is copied as is, while HS messages may need fragmentation */
+        if( cur->type == MBEDTLS_SSL_MSG_CHANGE_CIPHER_SPEC )
+        {
+            memcpy( ssl->out_msg, cur->p, cur->len );
+            ssl->out_msglen = cur->len;
+            ssl->out_msgtype = cur->type;
 
-        ssl->handshake->cur_msg = cur->next;
+            /* Update position inside current message */
+            ssl->handshake->cur_msg_p += cur->len;
+        }
+        else
+        {
+            const int ret_payload = mbedtls_ssl_get_max_out_record_payload( ssl );
+            const size_t max_record_payload = (size_t) ret_payload;
```

Comment:
`max_record_payload` seems redundant (I think I have already commented that, but it may have got lost during some history rewriting).

Reply:
I disagree - I think it's easier to read this way when distinguishing between the error code and the size value extracted from it, and the compiler will optimize it anyway. Also, in the follow-up PR #1918, this slightly changes, and the size extracted from the return value is needed more than once.

Reply:
Of well, not removing after all, see above.